src/boot/linux.c

```diff
@@ -119,16 +46,58 @@ EFI_STATUS linux_exec(
         if (err != EFI_SUCCESS)
                 return log_error_status(err, "Bad kernel image: %m");
 
-        _cleanup_(unload_imagep) EFI_HANDLE kernel_image = NULL;
-        err = load_image(parent, kernel->iov_base, kernel->iov_len, &kernel_image);
+        const PeSectionHeader *headers;
+        size_t n_headers;
+
+        /* Do we need to validate anyting here? the len? */
+        err = pe_section_table_from_base(kernel->iov_base, &headers, &n_headers);
         if (err != EFI_SUCCESS)
-                return log_error_status(err, "Error loading kernel image: %m");
+                return log_error_status(err, "Cannot read sections: %m");
+
+        /* Do we need to ensure under 4gb address on x86? */
+        _cleanup_pages_ Pages loaded_kernel_pages = xmalloc_pages(
+                        AllocateAnyPages, EfiLoaderCode, EFI_SIZE_TO_PAGES(kernel_size_in_memory), 0);
+
+        uint8_t* loaded_kernel = (uint8_t*)PHYSICAL_ADDRESS_TO_POINTER(loaded_kernel_pages.addr);
+        FOREACH_ARRAY(h, headers, n_headers) {
+                if (h->PointerToRelocations != 0)
+                        return log_error_status(EFI_LOAD_ERROR, "Inner kernel image contains sections with relocations, which we do not support.");
+                if (h->SizeOfRawData != 0) {
```

Comment:
i'd invert this check, and use `continue`so that indentation can be made more shallow below